source/DmapSupport.cpp

```diff
@@ -184,37 +184,51 @@ HRESULT GetControllerBaseAddress(PWCHAR deviceName, HANDLE & handle, PVOID & bas
                         g_openDeviceMask |= 1 << i;
 
                         IOControlCode^ IOCTL = ref new IOControlCode(0x423, 0x100, IOControlAccessMode::Any, IOControlBufferingMethod::Buffered);
-                        Buffer^ addressBuffer = ref new Buffer(8);
+                        Buffer^ addressBuffer = ref new Buffer(sizeof(DMAP_MAPMEMORY_OUTPUT_BUFFER));
 
                         create_task(device->SendIOControlAsync(IOCTL, nullptr, addressBuffer))
                             .then([addressBuffer, &controllerAddress, &findCompleted, &hr](UINT32 result)
                         {
-                            // We expect a 4-byte address and a 4-byte length to have been transferred 
-                            // into the address buffer by the I/O operation.
+                            // We expect an 8-byte address and a 4-byte length to have been transferred 
+                            // into the address buffer by the I/O operation.  If this is an old driver, 
+                            // we could get a 4-byte address and a 4-byte length instead.
 
-                            // hr should already be S_OK 
-                            if (result == 8)
+                            if (result < 8)
                             {
-                                hr = S_OK;
+                                hr = E_FAIL;
                             }
-
-                            if (SUCCEEDED(hr))
+                            else
                             {
                                 auto reader = DataReader::FromBuffer(addressBuffer);
-                                UINT32 address = { 0 };
+                                uint64_t address = 0;
                                 int i;
-                                for (i = 0; i < 4; i++)
+
+                                if (result < sizeof(DMAP_MAPMEMORY_OUTPUT_BUFFER))
                                 {
-                                    address = address | (((UINT32)reader->ReadByte()) << (8 * i));
+                                    // If we are talking to an old driver, parse a 4-byte address.
+                                    for (i = 0; i < 4; i++)
```

Comment:
Not really.  This check is designed to handle a situation in which the driver and the user-mode code disagree on what size the pointer field should be.  The buffer sent in is large enough for a struct with a 64-bit pointer (regardless of the size of int), but if the driver succeeds, but writes fewer bytes that the new struct, it is parsed as an old struct.  This should only happen when new user-mode code is run with an old driver.